src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+    fn into(self) -> liquid::value::Object {
+        let mut object = liquid::value::Object::new();
+        // if no pages, means we have indexes instead, `tags` like cases for exemple
+        if let Some(pages) = self.pages {
+            object.insert("pages".into(), liquid::value::Value::Array(pages));
+        }
+        // list of available indexes, in `tags` like cases
+        if let Some(indexes) = self.indexes {
+            object.insert(
+                "indexes".into(),
+                liquid::value::Value::Array(
+                    indexes.iter().map(liquid::value::Value::scalar).collect(),
+                ),
+            );
+        }
+        object.insert(
+            "index".into(),
+            liquid::value::Value::scalar(self.index as i32),
+        );
+        object.insert(
+            "index_permalink".into(),
+            liquid::value::Value::scalar(self.index_permalink),
+        );
+        if let Some(previous_index_permalink) = self.previous_index_permalink {
+            object.insert(
+                "previous_index".into(),
+                liquid::value::Value::scalar(self.previous_index as i32),
+            );
+            object.insert(
+                "previous_index_permalink".into(),
+                liquid::value::Value::scalar(previous_index_permalink),
+            );
+        }
+        if let Some(next_index_permalink) = self.next_index_permalink {
+            object.insert(
+                "next_index".into(),
+                liquid::value::Value::scalar(self.next_index as i32),
+            );
+            object.insert(
+                "next_index_permalink".into(),
+                liquid::value::Value::scalar(next_index_permalink),
+            );
+        }
+        object.insert(
+            "first_index_permalink".into(),
+            liquid::value::Value::scalar(self.first_index_permalink),
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc.front.pagination.as_ref().unwrap();
+    let mut all_posts = posts_data.to_vec();
+    match config.include {
+        Include::All => {
+            sort_posts(&mut all_posts, &config);
+            create_all_paginators(&mut all_posts, &doc, &config)
+        }
+        Include::None => Ok(Vec::new()), // user can set `include: None`
+    }
+}
+
+fn extract_value(a: &liquid::value::Value, key: &str) -> Option<liquid::value::Scalar> {
+    a.get(&key.into())
```

Comment:
What is the `into` doing?

Reply:
Transforms `key` into a `String`, otherwise:
```
error[E0277]: the trait bound `&liquid::value::Index: std::convert::From<&str>` is not satisfied
   --> src\pagination.rs:180:7
    |
180 |     a.get(key)
    |       ^^^ the trait `std::convert::From<&str>` is not implemented for `&liquid::value::Index`
    |
    = help: the following implementations were found:
              <liquid::value::Index as std::convert::From<isize>>
              <liquid::value::Index as std::convert::From<&'a str>>
              <liquid::value::Index as std::convert::From<std::string::String>>
    = note: required because of the requirements on the impl of `std::convert::Into<&liquid::value::Index>` for `&str`
```